R/simulateGLV.R

```diff
@@ -7,120 +7,213 @@
+    error_variance = 0,
+    norm = FALSE,
+    t_end = 1000, ...){
+
+    # set the default values
+    if (is.null(names_species)) {
+        names_species <- paste0("sp", seq_len(n_species))
+    }
+    if (is.null(A)) {
+        A <- randomA(n_species)
+    }
+    if (is.null(x0)) {
+        x0 <- runif(n = n_species, min = 0, max = 1)
+    }
+    if (is.null(growth_rates)) {
+        growth_rates <- runif(n = n_species, min = 0, max = 1)
     }
+    if (is.null(metacommunity_probability)) {
+        metacommunity_probability <- rdirichlet(1, alpha = rep(1,n_species))
+    }
+
+    #normalize metacommunity_probability
+    metacommunity_probability <- metacommunity_probability/
+        sum(metacommunity_probability)
+
+    # select the time points to simulate and to store
+    t_dyn <- simulationTimes(t_end = t_end, ...)
+
+    # calculate the time points influenced by the disturbances
+    tEvent <- eventTimes(
+        t_events = t_external_events,
+        t_duration = t_external_durations,
+        t_end = t_end,
+        ... = ...)
+
+    parameters <- list(growth_rates=growth_rates, A = A, n_species = n_species,
+        sigma_drift = sigma_drift, stochastic = stochastic,
+        sigma_epoch = sigma_epoch, epoch_p = epoch_p,
+        sigma_external = sigma_external, tEvent = tEvent,
+        migration_p = migration_p,
+        metacommunity_probability = metacommunity_probability,
+        sigma_migration= sigma_migration)
+
+    out <- ode(
+        y = x0,
+        times = t_dyn$t_sys,
+        func = glvModel,
+        parms = parameters,
+        events = list(func = perturb, time = t_dyn$t_sys),
+        maxsteps=10^9, method = "ode45")
+
+
+    out_matrix <- out[,2:ncol(out)]
+
+    out_matrix <- out_matrix[t_dyn$t_index,]
 
-dxdt <- function(t, x, parameters){
-    b <- parameters$b
-    A <- parameters$A
-    # rate of change
-    dx <- x*(b+A %*% x)
-    # return rate of change
-    list(dx)
+    if(error_variance > 0){
+        measurement_error <- rnorm(n = length(t_dyn$t_index)*n_species,
+                                   mean = 0, sd = sqrt(error_variance))
+        measurement_error <- matrix(measurement_error,
+                                    nrow = length(t_dyn$t_index))
+        out_matrix <- out_matrix + measurement_error
+    }
+
+    if(norm){
+        out_matrix <- out_matrix/rowSums(out_matrix)
+    }
+
+    colnames(out_matrix) <- names_species
+
+    out_matrix <- cbind(out_matrix, time = t_dyn$t_sys[t_dyn$t_index])
+
+    out_list <- list(matrix = out_matrix,
+        metacommunity_probability = metacommunity_probability,
+        migration_p = migration_p,
+        error_variance = error_variance)
+
+    return(out_list)
+}
+
+# define the GLV Model
+glvModel <- function(t, x0, parameters){
+  x0[x0 < 10^-8] <- 0
+  growth_rates <- parameters$growth_rates
+  A <- parameters$A
+  # rate of change
+  dx <- x0*(growth_rates+A %*% x0)
+  # return rate of change
+  list(dx)
```

Comment:
Indentation should be 4 spaces.

Reply:
all indentations are checked and re-styled using `styler::style_active_pkg()` with transformer set to `biocthis::bioc_style()`